fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +410,235 @@ impl ExecutorInner {
                 }
-            };
-            transitions = remaining_transitions;
-            let (transition_outcome, state, transition_fn, meta) = completed_result;
-            let operation_id = state.operation_id();
-            // info level because spans only log when any event happens within a span.
-            let span = tracing::info_span!("state_machine_transition", %operation_id);
-            async {
-                info!("Executing state transition");
-                debug!(
-                    ?state,
-                    transition_outcome = ?AbbreviateJson(&transition_outcome),
-                );
-
-                let active_or_inactive_state = self
-                    .db
-                    .autocommit(
-                        |dbtx, _| {
-                            let state = state.clone();
-                            let transition_fn = transition_fn.clone();
-                            let transition_outcome = transition_outcome.clone();
-                            Box::pin(async move {
-                                let new_state = transition_fn(
-                                    &mut ClientSMDatabaseTransaction::new(
-                                        &mut dbtx.to_ref(),
-                                        state.module_instance_id(),
-                                    ),
-                                    transition_outcome,
-                                    state.clone(),
-                                )
-                                .await;
-                                dbtx.remove_entry(&ActiveStateKey::from_state(state.clone()))
-                                    .await;
-                                dbtx.insert_entry(
-                                    &InactiveStateKey::from_state(state.clone()),
-                                    &meta.into_inactive(),
+                ExecutorLoopEvent::Triggered(TransitionForActiveState {
+                    outcome,
+                    state,
+                    meta,
+                    transition_fn,
+                }) => {
+                    let span = tracing::info_span!(
+                        "state_machine_transition",
+                        operation_id = %state.operation_id()
+                    );
+                    // Perform the transition as another future, so transitions can happen in
+                    // parallel.
+                    // Database write conflicts might be happening quite often here,
+                    // but transaction functions are supposed to be idempotent anyway,
+                    // so it seems like a good stress-test in the worst case.
+                    futures.push({
+                        let sm_update_tx = self.sm_update_tx.clone();
+                        let db = self.db.clone();
+                        let notifier = self.notifier.clone();
+                        let module_contexts = self
```

Comment:
The code creating this future is quite long and nested, did you think about extracting it into its own fn? Main challenge seems making it return a `'static` future while taking `&'a self`.

Reply:
The diff view (or any other width-limited view) would be easier to read :P